### ISSUE TITLE
pi07 ring attention (PR #298) produces NaN loss at step 2

### ISSUE THREAD

**Issue by shuheng-liu** [TensorAuto/OpenTau#300]
Tracking a NaN-loss reproduction observed while benchmarking [PR #298](https://github.com/TensorAuto/OpenTau/pull/298) under real-training conditions.

## Repro

- Branch: `feat/pi07-ring-attention` @ `f35c1fa`
- Config: copy of `configs/examples/pi07_pretrain_10percent_robotic_with_libero.json` (`type: pi07_low_level`), only `attention_implementation` flipped `sdpa → ring`, `batch_size: 16`, `log_freq: 1`
- Accelerate: ZeRO-2 + bf16, 8× A100 80GB, `gradient_accumulation_steps=1`
- `gradient_checkpointing: true` (config keeps it on; ignored under ring per PR #298)
- Pretrained init: `TensorAuto/pi07-low-untrained`

## What happened

```
step:1(1) smpl:128(128)   total_loss:18.524704 mse_loss:3.851591 ce_loss:14.673113 ... lr:2.5e-09 grad_norm:1.000
step:2(2) smpl:256(256)   total_loss:nan       mse_loss:nan       ce_loss:nan       ... lr:3.7e-09 grad_norm:1.000
```

Step 1 (pre-optimizer-step) is finite. Step 2 (first reported step after a parameter update) is fully NaN on every loss component. Both MSE and CE go NaN simultaneously, which suggests the corruption is in the parameters/activations rather than a single head.

## What I've ruled out

- **Not memory pressure**: cleanly within budget at 88% HBM (bs=16 → 72.3 GB worst-rank), no OOM.
- **Not a degenerate batch**: the L1/CE numbers at step 1 are normal-looking — bf16 didn't explode at construction time.
- **Not stochastic**: identical behaviour at bs=12 (saw `nan` at first log emit, step 100).
- **Not the action-tokenizer truncation warning**: the same warnings appear under SDPA without producing NaNs.

I have **not** verified yet:
- Whether the ring backward path has a bf16 overflow on the per-rank softmax-LSE accumulation (the obvious culprit if it's a numerics-of-ring issue).
- Whether disabling `gradient_checkpointing` in the config (it's ignored under ring anyway, but the wrapper still runs through the FSDP-conditional bf16 cast logic) changes anything.
- Whether the suffix forward (which routes through SDPA on full unsharded data) is the one going NaN, vs. the ring prefix.

## How to reproduce on an 8× A100 node

```bash
git checkout feat/pi07-ring-attention
uv sync --extra dev --extra libero
# Edit configs/examples/pi07_pretrain_10percent_robotic_with_libero.json:
#   "attention_implementation": "ring",
#   "log_freq": 1,
#   "batch_size": 16
accelerate launch \
  --config_file configs/examples/accelerate_deepspeed_config.yaml \
  --num_processes=8 \
  src/opentau/scripts/train.py \
  --config_path=configs/examples/pi07_pretrain_10percent_robotic_with_libero.json
```

The first two emitted steps reproduce the NaN.

## Related

The bench matrix on this PR (https://github.com/TensorAuto/OpenTau/pull/298#issuecomment-4436750223) used `profile_step.py`, which doesn't compute a loss — so it couldn't catch this. The unit/correctness tests in `src/opentau/scripts/ringattn_experiments/` check forward output drift only; they don't exercise the backward → optimizer step chain that explodes here.

**Comment by shuheng-liu** [comment]
Root cause localized via multi-stage probes injected at ring/gather backward boundaries and per-layer outputs.

## Probe results — backward gradient, iteration 1

Per-rank NaN/Inf counts at each backward boundary (in flow order, downstream → upstream):

| location | NaN | Inf | shape | what it is |
| --- | --- | --- | --- | --- |
| (just-after-gather, `final_norm` *output* grad) | **0** | **0** | (16, 2584, 2560) | gradient flowing INTO `final_norm.backward` |
| `final_norm` *input* grad = layer 14 output grad | ~2.9M (~22%) | **~657k (~5%)** | (16, 323, 2560) | gradient flowing OUT of `final_norm.backward` |
| layer 13 output grad | ~9.9M (~75%) | 0 | (16, 323, 2560) | one layer below |
| layer 12 output grad | ~12.4M (~94%) | 0 | (16, 323, 2560) | two layers below |
| ring `_RingAttention.backward.grad_O` entry (layer 14) | ~2.7M (~26%) | 0 | (16, 323, 8, 256) | the original symptom from the issue |

**`final_norm.backward` is where the corruption is born.** Input is clean; output has both Inf and NaN. Subsequent backward operations of the form `Inf * 0 = NaN` (residuals, layernorms, MLP backward) convert the Inf into NaN and the corruption spreads to nearly the entire activation tensor across layers.

## What's happening in the norm backward

The backbone's final norm is `gemma3.language_model.model.norm` — Gemma 3's stock RMSNorm. The forward is `x * rsqrt(mean(x^2) + eps) * (1 + weight)` with `eps=1e-6`. In bf16 mixed-precision the backward of `bf16_input * fp32_scalar` casts the gradient to fp32 internally, then back to bf16. When the rsqrt factor approaches its ceiling (`~1000` for `eps=1e-6` at near-zero variance), the multiplication `grad_y * rsqrt(...)` can push values past bf16's `3.4e38` ceiling → `+Inf` on cast. Once a single position lands on `+Inf`, subsequent backward ops produce `Inf * 0 = NaN` and the corruption cascades.

The reason this surfaces under **ring but not SDPA** appears to be the **grad_checkpointing asymmetry**: under SDPA, `gradient_checkpointing=True` is honored and activations are recomputed during backward — the gradient flowing into the final norm is computed against the *recomputed* layer outputs. Under ring, gradient_checkpointing is deliberately bypassed (per PR #298's design — ring's online softmax replaces the per-layer checkpoint) and the saved activations are used directly. Some interaction between the saved-vs-recomputed paths and the bf16 cast within Gemma3RMSNorm tips one over the bf16-overflow threshold and the other stays comfortably under.

## Reproduction methodology

Probes added as `_nan_probe` (printing NaN/Inf counts) at:
- `_RingAttention.forward.entry/exit` (Q/K/V, O, LSE)
- `_RingAttention.backward.entry/exit` (grad_O; dQ, dK, dV)
- `_AllGatherSeq.forward.entry/exit` and `.backward.entry/exit`
- `_GradProbe` autograd passthrough wrapping each layer's backbone output and the final norm's output

The asymmetry of probes firing exactly at `final_norm.backward.output` but not at `.input` is the key piece of evidence.

## Things I ruled out

- **Not gather_seq** — both the per-layer cache `gather_seq` (shape `(B, S_local, Hkv, Dh)`) and the final-output `gather_seq` (shape `(B, S_total, hidden)`) have clean forward and clean backward at iteration 1.
- **Not the ring kernel itself** — `_RingAttention.forward` returns clean O / LSE at iteration 1; `_RingAttention.backward` is just propagating the already-NaN `grad_O` it receives.
- **Not the CE / discrete-action head** — the gradient arriving at `final_norm.fwd_output` (before the norm's backward fires) is clean.
- **Not the tiny micro-repro** — a 2-layer 256-hidden synthetic ring forward + suffix-SDPA forward + sum-loss + backward + AdamW step iteration **PASSES** with no NaN. The bug needs the full pi07 stack (depth, hidden, real CE loss, untrained pretrained init).

## Recommended fix path for a follow-up PR

1. Replace the bf16-input × fp32-rsqrt multiply in the final norm path with a custom autograd Function that does both forward and backward purely in fp32, casting back to bf16 only after the gradient is fully computed (and clamping at bf16's representable range, not letting the cast silently overflow to ±Inf).
2. Or: drop the bypassed-grad-checkpointing assertion under ring and let `torch.utils.checkpoint` run anyway. Ring's blockwise rematerialisation handles attention activations, but it doesn't preempt the norm/MLP recomputation that the SDPA path gets for free. (The PR description argues checkpointing is "redundant" — the probe data suggests it isn't, at least for stabilising bf16 norm backward at this scale.)
3. Or: add `torch.nan_to_num(grad_O, nan=0.0, posinf=6e4, neginf=-6e4)` at `_RingAttention.backward` entry as a band-aid. This loses gradient at the affected positions (effectively zeroing CE-loss gradient on the discrete-action tail for the first step), but step 1 LR is in the 1e-9 warmup floor anyway so the lost signal is negligible. Not a real fix, but unblocks training.

## Status

Reverted the local checkout back to `main` and relaunched the prior SDPA training config so the cluster isn't sitting idle. Probe instrumentation is stashed locally on the branch.

**Comment by shuheng-liu** [comment]
## Follow-up: prior diagnosis ("bf16 cast overflow in `Gemma3RMSNorm.backward`") does not hold up

Picked this up to test the proposed fixes in the prior comment. Two negative results plus one open puzzle. Handing off — someone else can take it from here.

### 1. Path A (re-enable `gradient_checkpointing` under ring) does not fix the NaN

Deleted the `and not ring_active` clause in [`src/opentau/policies/pi07/gemma3_with_expert.py:1196`](https://github.com/TensorAuto/OpenTau/blob/feat/pi07-ring-attention/src/opentau/policies/pi07/gemma3_with_expert.py#L1196) so per-layer `torch.utils.checkpoint.checkpoint` runs under ring as well. (Required a second small fix: capture the resolved `attention_interface` once and pass it as an explicit positional argument to each layer call, otherwise the lazy `self._attention_interface_provider()` route reads `self._ring_active` at recompute time — by then the suffix forward has reset it to `False` and dispatch crashes with a shape mismatch when SDPA is invoked on sharded Q/K with a full-seq mask.)

With both changes applied:

```
step:1(1)   total_loss:18.524704   mse_loss:3.851591   ce_loss:14.673113   ...
step:2(2)   total_loss:nan         mse_loss:nan        ce_loss:nan         ...
```

Identical to baseline. The per-rank `layer14_backbone_output` backward NaN counts are bit-identical too (e.g. rank 0: `NaN=2944214 Inf=657706` with and without Path A). In hindsight this makes sense: the final `Gemma3RMSNorm` is applied **after** the layer loop in `Gemma3WithExpertModel.forward`, so it is not inside any `torch.utils.checkpoint` boundary regardless of `use_ckpt`. The "saved-vs-recomputed activation bf16 quantization" hypothesis doesn't apply at the final norm.

### 2. Custom-RMSNorm probe: the bf16 cast at the final norm is **not** where the Inf is born

Replaced `gemma3.language_model.model.norm.forward` in-place with a custom `torch.autograd.Function` whose forward replicates `Gemma3RMSNorm.forward` operation-by-operation and whose backward is hand-derived from the chain rule (same math PyTorch's autograd would generate). The backward prints `grad_y.absmax`, `inv.{max,min}`, `weight.absmax`, `x.{absmax,absmin}`, `grad_x_fp32.absmax`, and per-position NaN/Inf/overflow counts **before** the final `.to(grad_out.dtype)` cast back to bf16.

Iter-1 backward probe output on rank 0 (representative — all of ranks 0-6 look like this):

```
grad_y.absmax=0.000e+00      inv[max=6.36e-02, min=5.79e-02]
weight.absmax=5.35e+01       x[absmax=9.90e+01, absmin=0.000e+00]
grad_x_fp32.absmax=0.000e+00 fp32[nan=0, inf=0]
gy_nonzero_pos=0/5168        gx_nonzero_pos=0   gx_nan_pos=0   gx_overflow_pos=0
```

That is: the gradient entering the norm backward on the ranks whose post-gather chunk does not intersect the discrete-action slice is **exactly zero** (consistent with how the CE-loss gradient ends up entirely in the world-last rank's chunk after `gather_seq.backward`'s chunk-and-take), and the fp32 grad_x my hand-written backward produces is therefore also exactly zero. A `.to(bf16)` cast cannot overflow from a zero. A post-cast probe right before `return` confirms `grad_x_bf16.absmax=0, sum=0, nan=0, inf=0`.

Rank 7 (where the CE-loss gradient lands) is also clean — `grad_x_fp32.absmax=4.77e-05`, no overflow positions. The `(1 + weight) * inv * grad_y` amplification stays well under bf16's `3.4e38` ceiling for any realistic activation magnitude. The prior comment's "rsqrt-amplified gradient overflows on bf16 cast" mechanism does not match what the math actually computes here.

### 3. The actual corruption point (open question)

The downstream `_GradProbe('layer14_backbone_output').backward` still fires with `~22% NaN + 5% Inf` per rank in iter 1 — same counts as the baseline. Adding a finer probe at the entry to `_GradProbe.backward` (sample[0:8], `data_ptr`, and `torch.cuda.synchronize()` before `.item()`) shows it receives a tensor at **the same `data_ptr`** my custom backward just returned (e.g. `371319504384` on rank 0 in one run), but with the values now showing `absmax=nan, nan=2936533, inf=660267`. Between the two probes the memory at that address is modified in place.

Mechanisms I could not confirm or rule out from the data I have:
1. **Hidden additional autograd path** writing to `inputs_embeds[0]`'s grad accumulator. I searched the source and found only one Python-level consumer (`backbone_norm(hidden_states)` post-loop); if another consumer exists it isn't obvious from a read of the file.
2. **CUDA async-stream race** between ring's P2P send/recv buffers, DeepSpeed ZeRO-2's reduce-scatter, and the grad-accumulator memory. The probe with `cuda.synchronize()` before reads would have discriminated this but the GPU was reclaimed before iter-1 data landed.
3. **`_GradProbe.forward(ctx, x, tag): return x` returning its input unchanged** is non-recommended PyTorch usage; if autograd's view-tracking treats the wrapped output and its input as the same Variable, the accumulator behavior at the boundary could be surprising. Worth trying `return x.clone()` or `x.view_as(x)` as a discriminator.

If you're picking this up, the cheapest next experiment is (3): swap `_GradProbe.forward` to clone. If the iter-1 NaN counts shift, autograd view-tracking is involved. If not, eliminate that and chase (1) and (2).

### Adjacent finding (separate concern, worth its own issue)

Added a one-shot print of `prefix_embs[0, 0, :8]` per rank just before `split_seq(inputs_embeds[0], seq_dim=1)` at `gemma3_with_expert.py:1230`. On 8-rank DeepSpeed ZeRO-2 with `split_batches=False` (the production default — see `src/opentau/scripts/train.py:278`), all 8 ranks print **different** values. Each rank starts with its own data-parallel minibatch; `split_seq` shards that minibatch's sequence; `_AllGatherSeq.forward` then concatenates chunks from 8 different minibatches into a "Frankenstein" prefix on every rank. The CE loss reads `prefix_out[:, -151:-1]` (always inside the world-last rank's chunk), so each rank scores logits derived from the world-last rank's data against its own labels.

Result: a finite but semantically incoherent loss. This is independent of the iter-1 NaN — it would be wrong-by-construction even if the NaN fix landed. Likely needs either a sequence-parallel process group orthogonal to the DP group, or a broadcast of the minibatch across the ring group from a designated rank before sharding. I'll file separately.

### tl;dr

- Path A from the prior comment does nothing for this NaN.
- The norm-backward bf16-overflow mechanism described in the prior comment is contradicted by a bit-identical custom RMSNorm probe: fp32 `grad_x` is exactly zero (or 4.77e-05) where it needs to be, and no overflow occurs at the bf16 cast.
- The NaN/Inf at `layer14_backbone_output` is materializing **between** the norm backward's return and the next op's read of the same memory address. Mechanism unknown; needs deeper probing than I had time for.
- DDP × ring on the same WORLD process group produces a semantically broken `all_gather` regardless of the NaN — independent bug.
